tests/test_helpers.py

```diff
@@ -0,0 +1,33 @@
+from django.contrib.contenttypes.models import ContentType
+
+from djangocms_versioning.constants import DRAFT, PUBLISHED
+from djangocms_versioning.models import Version
+from filer.models import File, Folder
+
+from .base import BaseFilerVersioningTestCase
+from djangocms_versioning_filer import helpers
+
+
+class HelperTests(BaseFilerVersioningTestCase):
+
+    def test_check_folder_exists_in_folder(self):
+        self.assertFalse(helpers.check_folder_exists_in_folder(self.folder, 'olivertwist'))
+        subfolder = Folder(
+            parent=self.folder,
+            name='olivertwist'
+        )
+        subfolder.save()
+        self.assertEquals(helpers.check_folder_exists_in_folder(self.folder, 'olivertwist'), subfolder)
+
+    def test_add_subfolder(self):
+        self.assertFalse(helpers.check_folder_exists_in_folder(self.folder, 'olivertwist'))
+        subfolder = helpers.add_subfolder(self.folder, 'olivertwist')
```

Comment:
This is a unit test anti-pattern! Unless you want to write an integration test (which you shouldn't want to in this case :stuck_out_tongue: cause you're clearly checking the result of one small function) then you should never use project code to set up data for a test. This makes the test more annoying to maintain (if somebody changes/removes etc. `helpers.add_subfolder` they might end up having to fix this test which is completely unrelated to the functionality they're changing). And also, occasionally can cause really major issues (and people starting to distrust the test suite all together as a result) like tests passing when they should be failing (cause somebody who changes `helpers.add_subfolder` will have no idea it's being used in this test and after they change it, who knows if it still does what you assume it does in this test? Most of the time it still does, but I've seen the cases when it doesn't and they're really painful...).
When data for a test is tricky to set up I strongly advise using factories. Factory boy is currently the most popular python library for simplifying creation of test factories. Failing that a helper method on the test case class can help make the test a little more readable (but careful not to overuse that cause sometimes it can have the opposite effect).

Reply:
Hi Monika. 

I'm sorry but that does not compute. This test was designed to test the function `helpers.add_subfolder`. Your suggestion that the test should not use this function while testing this function makes no sense to me at all – it's a contradiction in terms is it not?

If someone changes `helpers.add_sub_folder` in such a way that breaks this test, then they will have broken both the function's intended purpose and the filer app...  and this test will have served its purpose.

Reply:
Ooops, sorry my bad - you're correct. I just got confused because I'm used to seeing AAA-like tests and my brain computed that as "data set up" for some reason. Apologies.